### PR TITLE
Multijobs travis builds

### PR DESCRIPTION
**Pull request by alaxalves**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -13,9 +13,10 @@
 # limitations under the License.
 
 
-gem "sinatra"
+gem "sinatra", "2.0.8.1"
```
Comment (alaxalves) [inline Gemfile:16]: Had to upgrade this, since https://github.com/sinatra/sinatra/pull/1533

**Comment by jywarren** [comment]
Actually I think I need to rely on @icarito for this one. Thanks!!

File: Dockerfile [hunk 2]
```diff
@@ -1,10 +1,24 @@
...
-FROM debian:buster
-
+FROM ruby:2.4.6-slim-buster as main-image
 MAINTAINER Sebastian Silva <sebastian@fuentelibre.org>
 
+FROM python:2.7 as gdal-builder
+
 # Install the application.
-RUN apt-get update -qq && apt-get install -y gdal-bin ruby imagemagick ruby-sinatra ruby-kramdown ruby-nokogiri bundler git python-gdal
+RUN apt-get update -qq && \
```
Comment (icarito) [inline Dockerfile:7]: What is the purpose of the changes to the Dockerfile?
It looks quite complex now, I was happy that it was pretty simple.
Also, I was using ruby-kramdown instead of pandoc which is way way smaller and doesn't pull all of build-essential compilers etc.
Please explain what is to be gained here.
Thanks!

File: scripts/install-deps.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+apt-get update -qq && \
+apt-get install -y \
+        gdal-bin \
+        python-gdal \
+        build-essential \
+        git \
```
Comment (icarito) [inline scripts/install-deps.sh:8]: Why do we need git and build-essential?

File: Dockerfile (hunk 2 above)
Comment (alaxalves) [reply Dockerfile:7]: I'm using the multistage build strategy on this Dockerfile, because I realized that the GDAL-related pkgs were always installing along python and some other python-related pkgs that already come with python's docker image.So I'm using the python image to install the python pkgs and copying where it has installed the images to our main ruby image. **FYI:** https://docs.docker.com/develop/develop-images/multistage-build/

PS: Ohh, I didn't realize this thing about pandoc. I'll switch back to ruby-kramdown.

File: scripts/install-deps.sh (hunk 3 above)
Comment (alaxalves) [reply scripts/install-deps.sh:8]: `build-essential` is needed for the `nokogiri` gem, since you were using the debian buster plain image, the C++ pkgs already came with it. The `git` pkg was already defined in the  Dockerfile [here](https://github.com/publiclab/mapknitter-exporter-sinatra/blob/main/Dockerfile#L7), so I just kept it.

Comment (icarito) [reply scripts/install-deps.sh:8]: Debian provides a nokogiri package which comes precompiled so there was no need to install compilers. It looks like you missed this dependency (in the same place as kramdown was) and that's why your build wants to compile its own nokogiri.
With regard to git, it's probably because `mapknitter-exporter` was originally imported from git before it was a published Gem. I think we could trim out Git.

File: Dockerfile (hunk 2 above)
Comment (icarito) [reply Dockerfile:7]: This is clever, but too clever in my humble opinion. What is the gain? Installed packages are kept low by avoiding compilers, is it really quicker?

Comment (icarito) [reply Dockerfile:7]: Also I fear blindly copying /usr/share and /usr/bin could introduce hard to debug issues. For instance, python packages often need stuff from /usr/lib. And there's no guarantee that ruby and python docker images are kept in sync so in principle there could be binary incompatibilities (for example, granted, unlikely in Debian Stable, but could happen, an updated glibc).

File: Dockerfile [hunk 4]
```diff
@@ -1,10 +1,24 @@
-# Debian base
-FROM debian:buster
-
+FROM ruby:2.4.6-slim-buster as main-image
 MAINTAINER Sebastian Silva <sebastian@fuentelibre.org>
 
+FROM python:2.7 as gdal-builder
```
Comment (icarito) [inline Dockerfile:4]: You see there's no guarantee that `python:2.7` will be based on Debian Buster in the future. If it is upgraded, this could all break. In any case it'd have to be 2.7-buster or similar and we would need to keep these in sync. Not sure if it's worth the complexity and risk.

Comment (alaxalves) [reply Dockerfile:4]: Thx for the feedback, I'll go back on onestage build as we're seeking for simplicity.

File: Dockerfile (hunk 2 above)
Comment (alaxalves) [reply Dockerfile:7]: The gain here was that we got to keep the image smaller (around ~300MB smaller). But since we're going for faster builds, then I'll go back to the onestage build. Thx for the input ;)

File: Dockerfile [hunk 5]
```diff
@@ -4,7 +4,17 @@ FROM debian:buster
 MAINTAINER Sebastian Silva <sebastian@fuentelibre.org>
 
 # Install the application.
-RUN apt-get update -qq && apt-get install -y gdal-bin ruby imagemagick ruby-sinatra ruby-kramdown ruby-nokogiri bundler git python-gdal
+RUN apt-get update -qq && \
+    apt-get install -y \
+                    gdal-bin \
+                    ruby \
+                    zlib1g-dev \
```
Comment (alaxalves) [inline Dockerfile:11]: Had to include this since: 
image
- https://travis-ci.com/github/publiclab/mapknitter-exporter-sinatra/jobs/319424879
- https://github.com/flapjack/omnibus-flapjack/issues/72

File: Gemfile [hunk 6]
```diff
@@ -13,9 +13,10 @@
...
 gem "sinatra-cors"
 gem "nokogiri"
+gem "pandoc-ruby"
```
Comment (alaxalves) [inline Gemfile:19]: Had to include this since:
image
- https://travis-ci.com/github/publiclab/mapknitter-exporter-sinatra/jobs/319423451

File: scripts/install-deps.sh [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+apt-get update -qq && \
+apt-get install -y \
+        gdal-bin \
+        python-gdal \
+        pandoc \
```
Comment (alaxalves) [inline scripts/install-deps.sh:7]: Had to include this since: 
image
- https://travis-ci.com/github/publiclab/mapknitter-exporter-sinatra/jobs/319424125

**Comment by alaxalves** [comment]
@icarito Could you review this again please?

File: scripts/install-deps.sh (hunk 7 above)
Comment (icarito) [reply scripts/install-deps.sh:7]: Hi,
In second review I still see that you've introduced `pandoc` as dependency. In previous version `kramdown` did the exact same and pandoc is just an optional dependency (which is very large). Please try again without including pandoc, I took care that it wasn't required, unless I am wrong. Thanks for the perseverance!
```suggestion
        pandoc \
```

File: Dockerfile (hunk 5 above)
Comment (icarito) [reply Dockerfile:11]: Nokogiri shouldn't need to be built since it's included as a debian package which comes already compiled.

File: scripts/install-deps.sh (hunk 7 above)
Comment (alaxalves) [reply scripts/install-deps.sh]: @icarito When I remove pandoc I get the error:
image

Via: https://travis-ci.com/github/publiclab/mapknitter-exporter-sinatra/jobs/345621753

Comment (alaxalves) [reply scripts/install-deps.sh]: Should I revert 3428c40075a3c12371b8948f76cf7192296e844c to get this merged?

**Comment by alaxalves** [comment]
Closing this since https://github.com/publiclab/mapknitter-exporter-sinatra/pull/58
